### PR TITLE
Add learning-rate parameter

### PR DESCRIPTION
**Pull request by Elhl93**
Add parameters for learning rate for solo.train and vae.

### CODE REVIEW

**Comment by njbernstein** [comment]
Great. Taking a look now!

File: solo/solo.py [hunk 1]
```diff
@@ -194,7 +210,7 @@ def main():
     batch_size = params.get("batch_size", 128)
     valid_pct = params.get("valid_pct", 0.1)
     check_val_every_n_epoch = params.get("check_val_every_n_epoch", 5)
-    learning_rate = params.get("learning_rate", 1e-3)
+    learning_rate = params.get("learning_rate", lr_st)
```
Comment (njbernstein) [inline solo/solo.py:213]: needs to be `args.lr_st`

File: solo/solo.py [hunk 2]
```diff
@@ -226,7 +242,7 @@ def main():
         plan_kwargs = {
             "reduce_lr_on_plateau": True,
             "lr_factor": 0.1,
-            "lr": 1e-2,
+            "lr": lr_vae,
```
Comment (njbernstein) [inline solo/solo.py:245]: needs to be `args. lr_vae`

File: solo/solo.py (hunk 1 above)
Comment (njbernstein) [reply solo/solo.py:213]: probably also rename learning_rate completely to `lr_st`

**Review by njbernstein — CHANGES_REQUESTED** [review]
Need the changes described. I'll then do performance testing

File: solo/solo.py [hunk 3]
```diff
@@ -226,7 +242,7 @@ def main():
         plan_kwargs = {
             "reduce_lr_on_plateau": True,
             "lr_factor": 0.1,
-            "lr": 1e-2,
+            "lr": args.lr_vae,
```
Comment (Elhl93) [inline solo/solo.py:245]: added args.

File: README.md [hunk 4]
```diff
@@ -53,6 +53,11 @@ optional arguments:
   -recalibrate_scores   Recalibrate doublet scores (not recommended anymore)
                         (default: False)
   --version             Get version of solo-sc (default: False)
+  --lr_st
```
Comment (Elhl93) [inline README.md:56]: changed name of parameter in function

File: solo/solo.py [hunk 5]
```diff
@@ -194,7 +210,7 @@ def main():
     batch_size = params.get("batch_size", 128)
     valid_pct = params.get("valid_pct", 0.1)
     check_val_every_n_epoch = params.get("check_val_every_n_epoch", 5)
-    learning_rate = params.get("learning_rate", 1e-3)
+    learning_rate = params.get("learning_rate", args.lr_st)
```
Comment (Elhl93) [inline solo/solo.py:213]: added args.

File: solo/solo.py [hunk 6]
```diff
@@ -123,6 +123,22 @@ def main():
         action="store_true",
         help="Get version of solo-sc",
     )
+
```
Comment (Elhl93) [inline solo/solo.py:126]: changed names to lr_st and lr_vae

File: README.md [hunk 7]
```diff
@@ -21,7 +21,7 @@ usage: solo [-h] -j MODEL_JSON_FILE -d DATA_PATH
             [--set-reproducible-seed REPRODUCIBLE_SEED]
             [--doublet-depth DOUBLET_DEPTH] [-g] [-a] [-o OUT_DIR]
             [-r DOUBLET_RATIO] [-s SEED] [-e EXPECTED_NUMBER_OF_DOUBLETS] [-p]
-            [-recalibrate_scores] [--version]
+            [-recalibrate_scores] [--version] [--lr_st] [--lr_vae]
```
Comment (Elhl93) [inline README.md:24]: changed names in README

File: README.md [hunk 8]
```diff
@@ -66,7 +71,7 @@ model_json example:
   "cl_hidden": 128,
   "cl_layers": 1,
   "dropout_rate": 0.2,
-  "learning_rate": 0.001,
+  "lr_st": 1e-5,
```
Comment (Elhl93) [inline README.md:74]: changed

File: solo/solo.py [hunk 9]
```diff
@@ -123,6 +123,22 @@ def main():
...
+        "--lr_st",
+        dest="lr_st",
+        default=1e-5,
+        type=int,
+        help="Learning rate used for solo.train",
+    )
+    parser.add_argument(
+        "--lr_vae",
+        dest="lr_vae",
+        default=1e-5,
```
Comment (njbernstein) [inline solo/solo.py:137]: The code all runs fine. On my bench mark datasets these defaults lead to worse performance. Can we change them to `1e-3`?

**Review by njbernstein — CHANGES_REQUESTED** [review]
Just the suggested change and then I will merge it in

File: solo/solo.py (hunk 9 above)
Comment (Elhl93) [reply solo/solo.py:137]: I only changed the lr_vae parameter and obtained the following results. Interestingly the results on training were slightly (0.01) lower, for validation results were slightly better. Good news: lr_vae=1e-3 ran through without any problems.

# LR 1e-3
Training results
AUROC: 0.96, Accuracy: 0.97, Average precision: 0.97
Validation results
AUROC: 0.90, Accuracy: 0.91, Average precision: 0.91

# LR 1e-5
AUROC: 0.96, Accuracy: 0.96, Average precision: 0.96
Validation results
AUROC: 0.92, Accuracy: 0.93, Average precision: 0.94

If it performs better on your benchmark dataset, I agree with you, lets go with 1e-3

**Review by Elhl93 — COMMENTED** [review]
Changing lr_st and lr_vae to 1e-3

**Review by Elhl93 — COMMENTED** [review]
Changed lr_st suggestion in README

File: README.md [hunk 10]
```diff
@@ -53,6 +53,11 @@ optional arguments:
   -recalibrate_scores   Recalibrate doublet scores (not recommended anymore)
                         (default: False)
   --version             Get version of solo-sc (default: False)
+  --lr_st            
+                        Learning rate used for solo.train (default: 1e-5)
+  --lr_vae             
+                        Learning rate used for vae (default: 1e-5)
```
Comment (njbernstein) [inline README.md:59]: change these to 1e-3 @Elhl93  please

**Review by Elhl93 — COMMENTED** [review]
Changed to 1e-3 in "How to Solo"

**Review by Elhl93 — COMMENTED** [review]
An additional line in the manual to emphasize switching to learning rate 1e-5 in case 1e-3 throws an error -

**Comment by Elhl93** [comment]
We just received another chunk of samples, unfortunately 3 samples (out of 50) return the same NaN error. 1e-5 they work fine. I added a note in the Readme so that people can change the value in case 1e-3 does not work at all

**Review by njbernstein — APPROVED** [review]
